src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -0,0 +1,1133 @@
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+static int
+_services_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+
+    pending->service_slot = sd_bus_slot_unref(pending->service_slot);
+
+    if (sol_bus_log_callback(m, userdata, ret_error) < 0)
+        return -EINVAL;
+
+    return services_list_changed(m);
+}
+
+static int
+_manager_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+    char *str;
+    int r;
+
+    pending->manager_slot = sd_bus_slot_unref(pending->manager_slot);
+
+    if (sol_bus_log_callback(m, userdata, ret_error) < 0)
+        return -EINVAL;
+
+    r = sd_bus_message_enter_container(m, SD_BUS_TYPE_ARRAY, "{sv}");
+    SOL_INT_CHECK(r, < 0, r);
+
+    do {
+
+        r = sd_bus_message_enter_container(m, SD_BUS_TYPE_DICT_ENTRY, "sv");
+        SOL_INT_CHECK_GOTO(r, < 1, end);
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+
+        if (streq(str, "State")) {
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else {
+            SOL_DBG("Ignored global manager property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0)
+        r = sd_bus_message_exit_container(m);
+    else
+        return r;
+
+    call_manager_monitor_callback();
+
+    return r;
+}
+
+static int
+dbus_connection_get_manager_properties(void)
+{
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    return sd_bus_call_method_async(bus, &_ctx.manager_slot,
+        "net.connman", "/", "net.connman.Manager", "GetProperties",
+        _manager_properties_changed, &_ctx, NULL);
+}
+
+static int
+dbus_connection_get_service_properties(void)
+{
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    return sd_bus_call_method_async(bus, &_ctx.service_slot,
+        "net.connman", "/", "net.connman.Manager", "GetServices",
+        _services_properties_changed, &_ctx, NULL);
+}
+
+SOL_API enum sol_connman_state
+sol_connman_get_state(void)
+{
+    return _ctx.connman_state;
+}
+
+static int
+_set_state_property_changed(sd_bus_message *reply, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+
+    pending->state_slot = sd_bus_slot_unref(pending->state_slot);
+
+    return sol_bus_log_callback(reply, userdata, ret_error);
+}
+
+SOL_API int
+sol_connman_set_radios_offline(bool enabled)
+{
+    int r;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sd_bus_call_method_async(bus, &_ctx.state_slot, "net.connman", "/",
```

Comment:
what happens if there was an ongoing call? will `state_slot` leak? Should you use a ptr_vector instead?